pulumi/grapl/__main__.py

```diff
@@ -66,6 +77,104 @@ def grapl_core_docker_image_tags(
+# inputs using Python classes, this does not appear to work (again,
+# due to serialization issues). Instead, we can use plain Python
+# dictionaries. Not ideal, but it has the benefit of actually working.
+class ConsulAclBootstrap(Resource):
+    id: pulumi.Output[str]
+    consul_address: pulumi.Output[str]
+    secret_token: pulumi.Output[str]
+
+    def __init__(
+        self,
+        name: str,
+        consul_address: str,
+        secret_token: Optional[str] = None,
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ):
+
+        super().__init__(
+            ConsulAclBootstrapProvider(),
+            name,
+            {
+                "consul_address": consul_address,
+                "secret_token": secret_token,
+            },
+            opts,
+        )
+
+
+class ConsulAclBootstrapProvider(ResourceProvider):
+    # The consul acl bootstrapping can only be run once, after that it returns a 403.
+    # If this is running on a previously bootstrapped cluster, we require a token be configured and create a fake id
+    def create(self, inputs: Mapping[str, Any]) -> CreateResult:
+        response = requests.put(f"{inputs['consul_address']}/v1/acl/bootstrap")
+        if response.status_code == requests.codes.ok:
+            bootstrap_id = response.json()["AccessorID"]
+            secret_id = response.json()["SecretID"]
+        elif response.status_code == requests.codes.forbidden:
+            # We've already run the bootstrap process so instead let's grab the token from the config
+            if inputs["secret_token"] is not None:
+                secret_id = inputs["secret_token"]
+            else:
+                raise Exception(
+                    "If Consul ACL Bootstrapping is complete you must set consul:token with the token"
+                )
+            # Create a fake static id since we don't have the accessor_id available
+            bootstrap_id = "fake-id-since-previously-bootstrapped"
+        else:
+            response.raise_for_status()
+
+        # As is customary in Pulumi, all inputs are available as
+        # outputs (this must include our injected Vault address and
+        # namespace values, which will be required when we try to
+        # delete the resource).
+        #
+        outs = cast(MutableMapping[str, Any], deepcopy(inputs))
+        # ... plus the secret token
+        outs["id"] = bootstrap_id
+        outs["secret_token"] = secret_id
+        return CreateResult(id_=bootstrap_id, outs=outs)
+
+    def delete(self, id: str, props: Mapping[str, Any]) -> None:
+        pass
+
+    # The function that determines if an existing resource whose inputs were
+    # modified needs to be updated or entirely replaced
+    def diff(
+        self, id: str, old_inputs: Mapping[str, Any], new_inputs: Mapping[str, Any]
+    ) -> DiffResult:
+        """
+        This is a no-op since consul acl bootstrapping can only be run once
```

Comment:
If it's a no-op, is a `diff` implementation needed?

Reply:
You're correct this is unnecessary